### PR TITLE
Sanitize number of radical electrons in case of invalid value assigned by CXSMILES

### LINKED ISSUE #6370: non-physical radical counts being preserved
**Describe the bug**
Sanitization is leaving radicals on molecules where they don't make sense.

**To Reproduce**
```
In [8]: Chem.MolFromSmiles('[NH4+] |^1:0|').Debug()
Atoms:
	0 7 N chg: 1  deg: 0 exp: 4 imp: 0 hyb: 4 arom?: 0 chi: 0 rad: 1
Bonds:

In [9]: Chem.MolFromSmiles('[CH4] |^1:0|').Debug()
Atoms:
	0 6 C chg: 0  deg: 0 exp: 4 imp: 0 hyb: 4 arom?: 0 chi: 0 rad: 1
Bonds:

```
Neither of those should have a radical on it.

**Configuration (please complete the following information):**
 - RDKit version: all
 - OS: all

### PR DESCRIPTION
**Pull request by MarioAndWario**
#### Reference Issue
Fixes #6370 

#### What does this implement/fix? Explain your changes.
Number of radical electrons are first assigned by the regular SMILES, and then get overridden by the CXSMILES extension. For example,

```python
Chem.MolFromSmiles('[CH4]').Debug()
#	0 6 C chg: 0  deg: 0 exp: 4 imp: 0 hyb: 4 arom?: 0 chi: 0
Chem.MolFromSmiles('[CH4] |^1:0|').Debug()
#	0 6 C chg: 0  deg: 0 exp: 4 imp: 0 hyb: 4 arom?: 0 chi: 0 rad: 1
Chem.MolFromSmiles('[CH4] |^2:0|').Debug()
#	0 6 C chg: 0  deg: 0 exp: 4 imp: 0 hyb: 4 arom?: 0 chi: 0 rad: 2
```

The radical info from CXSMILES may be invalid and therefore should be sanitized. However, current implementation of `assignRadicals()` within `sanitizeMol()` will do nothing if the preset number of radical electrons is non-zero. The fix is then to allow such sanitization by removing the condition `if (atom->getNumRadicalElectrons())`.

Unittests are added for a set of CXSMILES:
1. `[XHn] |^m:0|` where X is the element symbol, n = 1, ..., valence, m = 1, ..., 7 denotes the radical type.
2. `[NH4+] |^m:0|` where m = 1, ..., 7 denotes the radical type.

I also fix a minor issue in `calcImplicitValence()` where `d_implicitValence` is not overwritten by the return value.

### CODE REVIEW

**Comment by MarioAndWario** [comment]
1. The test failures in `testMolWrite` and `fileParsersTest1` are caused by incorrect radical type assigned to the 1st C in both `[C]C` and `[CH]`. The correct radical type should be trivalent, while the unittests expect monovalent.
```python
Chem.MolFromSmiles('[C]C').Debug()
# 0 6 C chg: 0  deg: 1 exp: 1 imp: 0 hyb: 4 arom?: 0 chi: 0 rad: 3
# 1 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
Chem.MolFromSmiles('[CH]').Debug()
# 0 6 C chg: 0  deg: 0 exp: 1 imp: 0 hyb: 4 arom?: 0 chi: 0 rad: 3
```
2. The test failure in `testMrvToMol` is caused by the incorrect radical type assigned to B in `[B-]C#N`.
```python
Chem.MolFromSmiles("[B-]C#N").Debug()
# 0 5 B chg: -1  deg: 1 exp: 1 imp: 0 hyb: 4 arom?: 0 chi: 0 rad: 3
# 1 6 C chg: 0  deg: 2 exp: 4 imp: 0 hyb: 2 arom?: 0 chi: 0
# 2 7 N chg: 0  deg: 1 exp: 3 imp: 0 hyb: 2 arom?: 0 chi: 0
```
Here B should be a trivalent radical center but the `.mrv` [file](https://github.com/rdkit/rdkit/blob/4a69bc3493dd3e9bb9f7a519e306fbcb545f1452/Code/GraphMol/MarvinParse/test_data/mrvValence_value.mrv.expected.mrv#L2) expects a monovalent one.

**Comment by greglandrum** [comment]
@MarioAndWario I'm very sorry for how long it took for me to review this.

File: Code/GraphMol/test1.cpp [hunk 1]
```diff
@@ -1574,6 +1575,60 @@ void testGithub1843() {
   BOOST_LOG(rdErrorLog) << "Finished" << std::endl;
 }
 
+void testGithub6370() {
```
Comment (greglandrum) [inline Code/GraphMol/test1.cpp:1578]: Nice. This is a really comprehensive test

File: Code/GraphMol/FileParsers/test1.cpp [hunk 2]
```diff
@@ -2474,12 +2474,12 @@ void test3V3K() {
...
     // radical and valence
     fName = rdbase + "CH.v3k.mol";
-    RWMol *m = MolFileToMol(fName);
+    RWMol *m = MolFileToMol(fName);  // cxsmiles = "[CH] |(0,0,),^5:0|"
     TEST_ASSERT(m);
     TEST_ASSERT(m->getNumAtoms() == 1);
     TEST_ASSERT(m->getAtomWithIdx(0)->getNoImplicit());
     TEST_ASSERT(m->getAtomWithIdx(0)->getNumExplicitHs() == 1);
-    TEST_ASSERT(m->getAtomWithIdx(0)->getNumRadicalElectrons() == 1);
+    TEST_ASSERT(m->getAtomWithIdx(0)->getNumRadicalElectrons() == 3);
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/FileParsers/test1.cpp:2482]: My suggestion would be to add something like the following to the "Backwards Incompatible Changes" section of the release notes:

```
  - The way that the number of radical electrons is calculated for atoms coming from mol blocks has been changed. Systems like a `[CH]` marked as a `DOUBLET` will now have three radical electrons assigned. This is consistent with the value from SMILES.
```

**Review by greglandrum — CHANGES_REQUESTED** [review]
Generally looks good. I made a couple of small suggestions.

File: Code/GraphMol/test1.cpp [hunk 3]
```diff
@@ -14,14 +14,15 @@
 #include <GraphMol/RDKitQueries.h>
 #include <RDGeneral/types.h>
 #include <RDGeneral/RDLog.h>
-//#include <boost/log/functions.hpp>
+// #include <boost/log/functions.hpp>
```
[the remark this answers is not in the document]
Comment (MarioAndWario) [reply Code/GraphMol/test1.cpp]: Deleted

File: Code/GraphMol/FileParsers/test1.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarioAndWario) [reply Code/GraphMol/FileParsers/test1.cpp:2483]: Added.

File: Code/GraphMol/MolOps.cpp [hunk 4]
```diff
@@ -388,8 +388,7 @@ void assignRadicals(RWMol &mol) {
   for (auto atom : mol.atoms()) {
     // we only put automatically assign radicals to things that
     // don't have them already and don't have implicit Hs:
```
[the remark this answers is not in the document]
Comment (MarioAndWario) [reply Code/GraphMol/MolOps.cpp]: Removed

**Comment by MarioAndWario** [comment]
> @MarioAndWario I'm very sorry for how long it took for me to review this.

No worries! I have applied your suggested changes in the latest commit.

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
@MarioAndWario thanks for the contribution!

**Pull request merged**